service/raft/raft_group0.cc

```diff
@@ -102,7 +102,10 @@ raft_server_for_group raft_group0::create_server_for_group0(raft::group_id gid,
     auto storage = std::make_unique<raft_sys_table_storage>(_qp, gid, my_addr.id);
     auto& persistence_ref = *storage;
     auto server = raft::create_server(my_addr.id, std::move(rpc), std::move(state_machine),
-            std::move(storage), _raft_gr.failure_detector(), raft::server::configuration());
+            std::move(storage), _raft_gr.failure_detector(),
+            raft::server::configuration {
+                .on_background_error = [](std::exception_ptr e) { on_internal_error(rslog, e); }
```

Comment:
`on_internal_error` throws by default, so how it improves the situations?

Reply:
This enables a standard internal error handling scheme. Why it throws by default is another question, I would
prefer an abort, as broken invariants could cause unpredictable system behaviour.

Reply:
There is no "standard internal error handling scheme". Where there is a broken invariant `abort` is used directly, Where an application may continue despite of unexpected error `on_internal_error` is used. It throws in production, so that a node will continue functioning, but aborts in testing to help finding errors. Here we want a node to stop gracefully if group0 cannot function. This is neither `abort` nor `on_internal_error`.

The patch in production is noop as it is now.

Reply:
@avikivity maybe we could change the default? `on_internal_error` means we’ve entered some unexpected state an it could cause unpredictable consequences and damage.

Reply:
> There is no "standard internal error handling scheme". Where there is a broken invariant `abort` is used directly, Where an application may continue despite of unexpected error `on_internal_error` is used. It throws in production, so that a node will continue functioning, but aborts in testing to help finding errors. Here we want a node to stop gracefully if group0 cannot function. This is neither `abort` nor `on_internal_error`.
> 
> The patch in production is noop as it is now.

Why this is not an abort? There should be no unhandled exceptions in io_fiber, if get some some it means something should be fixed in code.

Reply:
The exception is not unhandled. It is stored in `_io_status` and handled during `server::stop()`. We want this exception to notify a server's creator to do something about it. Either shutdown the node or restart the group. `on_internal_error` does neither.

Reply:
> The exception is not unhandled. It is stored in _io_status and handled during server::stop().
No, [it isn't](https://github.com/scylladb/scylladb/pull/11332/files#diff-7c8eecebd17f6e9f8791f2e135952aa7e6b39430d95816a30325926c3692b4c8L933), it's get caught and logged without rethrow. I'll replace `on_internal_error` with `request_abort` in `on_background_error`.

Reply:
> > The exception is not unhandled. It is stored in _io_status and handled during server::stop().
> 
> No, [it isn't](https://github.com/scylladb/scylladb/pull/11332/files#diff-7c8eecebd17f6e9f8791f2e135952aa7e6b39430d95816a30325926c3692b4c8L933), it's get caught and logged without rethrow. I'll replace `on_internal_error` with `request_abort` in `on_background_error`.

Yes, it was ignored here before your patch. I am talking about the code after your patch. It rethrowns and the rethrown exception is stored in `_io_status`. It is ignored in server::abort() instead.